aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -345,17 +345,13 @@ Tensor _prod(const Tensor &self, int64_t dim_, bool keepdim) {
 
 Tensor& logsumexp_out(Tensor& result, const Tensor &self, int64_t dim_, bool keepdim) {
   int64_t dim = maybe_wrap_dim(dim_, self.dim());
-  // can't take max of empty tensor.
-  if (self.numel() != 0) {
-    auto maxes = at::max_values(self, dim, true);
-    result = at::where((maxes == INFINITY).__or__(maxes == -INFINITY),
-                       maxes,
-                       maxes + at::log(at::sum(at::exp(self - maxes), dim, true)));
-  } else {
-    result = at::log(at::sum(at::exp(self), dim, true));
-  }
-  if (! keepdim)
-    result.squeeze_(dim);
+  // can't take max of empty tensor
+  AT_CHECK(self.numel() != 0, "logsumexp only works on nonempty tensors");
+  auto maxes = at::max_values(self, dim, true);
+  auto maxes_squeezed = (keepdim ? maxes : maxes.squeeze(dim));
+  maxes_squeezed.masked_scatter_(maxes_squeezed.abs() == INFINITY, at::zeros({1}, maxes_squeezed.options()).expand_as(maxes_squeezed));
```

Comment:
Is that last `expand_as` necessary? Shouldn't `maxes_squeezed` already be the correct size?

Reply:
It's to get the zeros to the size of maxes_squeezed. masked_scatter_ does not broadcast, apparently.